pkg/snet/stun_handler.go

```diff
@@ -0,0 +1,420 @@
+		retransmissionTimer.rto = 500 * time.Millisecond
+		retransmissionTimer.srtt = 0
+		retransmissionTimer.rttvar = 0
+	}
+
+	c.stunChans[txID] = make(chan stunResponse, Rc*2)
+
+	var ctx context.Context
+	var cancel context.CancelFunc
+
+	if c.writeDeadline.IsZero() {
+		ctx, cancel = context.WithCancel(context.Background())
+	} else {
+		ctx, cancel = context.WithDeadline(context.Background(), c.writeDeadline)
+	}
+	g, ctx := errgroup.WithContext(ctx)
+
+	c.mutex.Unlock()
+
+	defer func() {
+		c.mutex.Lock()
+		delete(c.stunChans, txID)
+		c.mutex.Unlock()
+	}()
+
+	isRetransmission := atomic.Bool{}
+	isRetransmission.Store(false)
+
+	var mappedAddress netip.AddrPort
+	var startTime, endTime time.Time
+
+	// Sending goroutine
+	g.Go(func() error {
+		originalRTO := retransmissionTimer.rto
+		currentRTO := originalRTO
+		startTime = time.Now()
+		for i := 0; i < Rc; i++ {
+			_, err := c.WriteTo(stunRequest, dest)
+			if err != nil {
+				return err
+			}
+
+			if i == 1 {
+				isRetransmission.Store(true)
+			}
+
+			var waitDuration time.Duration
+			if i < Rc-1 {
+				waitDuration = currentRTO
+				currentRTO *= 2
+			} else {
+				waitDuration = Rm * originalRTO
+			}
+
+			select {
+			case <-time.After(waitDuration):
+				// Continue to next iteration or timeout
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+		return context.DeadlineExceeded
+	})
+
+	// Receiving goroutine
+	g.Go(func() error {
+		var err error
+		mappedAddress, err = c.getStunResponse(ctx, txID)
+		cancel()
+		if err != nil {
+			return err
+		}
+		endTime = time.Now()
+		return nil
+	})
+
+	if err := g.Wait(); err != nil && !errors.Is(err, context.Canceled) {
+		return nil, err
+	}
+
+	mappedAddr, err := net.ResolveUDPAddr("udp", mappedAddress.String())
+	if err != nil {
+		return nil, err
+	}
+
+	c.mutex.Lock()
+	mapping := c.mappings[dest]
+	if mapping == nil {
+		mapping = &natMapping{destination: dest}
+		c.mappings[dest] = mapping
+	}
+	mapping.mappedAddr = mappedAddr
+	mapping.touch()
+
+	// Skip RTT calculation on retransmission or error
+	if isRetransmission.Load() || endTime.IsZero() {
+		c.mutex.Unlock()
+		return mapping, nil
+	}
+
+	// Update retransmission timer based on measured RTT, see RFC 6298
+	rtt := endTime.Sub(startTime)
+	if retransmissionTimer.srtt == 0 {
+		retransmissionTimer.srtt = rtt
+		retransmissionTimer.rttvar = rtt / 2
+	} else {
+		srttDiff := retransmissionTimer.srtt - rtt
+		if srttDiff < 0 {
+			srttDiff = -srttDiff
+		}
+		retransmissionTimer.rttvar = (3*retransmissionTimer.rttvar + srttDiff) / 4
+		retransmissionTimer.srtt = (7*retransmissionTimer.srtt + rtt) / 8
+	}
+	maxTerm := retransmissionTimer.rttvar * 4
+	if maxTerm < time.Millisecond {
+		maxTerm = time.Millisecond
+	}
+	retransmissionTimer.rto = retransmissionTimer.srtt + maxTerm
+	retransmissionTimer.lastUsed = time.Now()
+	c.mutex.Unlock()
+	return mapping, nil
+}
+
+func (c *stunHandler) SetWriteDeadline(t time.Time) error {
```

Comment:
I wonder what are the semantics for this function. My guess is: "This function sets the WriteDeadline for the underlay socket and any context associated with the execution flow for the NAT resolution?". What's the reasoning for doing this only for `SetWriteDeadline()` and not `SetDeadline()` and/or `SetReadDeadline()`.